packages/client/utils/AzureDevopsClientManager.ts

```diff
@@ -0,0 +1,346 @@
+      //     teamId
+      //   },
+      //   {onError, onCompleted}
+      // )
+      popup && popup.close()
+      window.removeEventListener('message', handler)
+    }
+    window.addEventListener('message', handler)
+  }
+
+  accessToken: string
+  refreshToken?: string
+  private readonly get: (url: string) => any
+  private readonly post: (url: string, payload: object) => any
+  // the any is for node until we can use tsc in nodeland
+  cache: {[key: string]: {result: any; expiration: number | any}} = {}
+  timeout = 5000
+
+  constructor(accessToken: string, options: AzureDevopsClientManagerOptions = {}) {
+    this.accessToken = accessToken
+    this.refreshToken = options.refreshToken
+    const fetch = options.fetch || window.fetch
+    const headers = {
+      // an Authorization requires a preflight request, ie reqs are slow
+      Authorization: `Bearer ${accessToken}`,
+      Accept: 'application/json' as 'application/json',
+      'Content-Type': 'application/json'
+    }
+    this.post = async (url, payload) => {
+      const res = await fetch(url, {
+        method: 'POST',
+        headers,
+        body: JSON.stringify(payload)
+      })
+      return res.json()
+    }
+
+    this.get = async (url) => {
+      const record = this.cache[url]
+      if (!record) {
+        const res = await fetch(url, {headers})
+        const result = await res.json()
+        this.cache[url] = {
+          result,
+          expiration: setTimeout(() => {
+            delete this.cache[url]
+          }, this.timeout)
+        }
+      } else {
+        clearTimeout(record.expiration)
+        record.expiration = setTimeout(() => {
+          delete this.cache[url]
+        }, this.timeout)
+      }
+      return this.cache[url].result
+    }
+  }
+
+  async getAzureDevOpsAccounts() {
+    return this.get('https://app.vssps.visualstudio.com/_apis/accounts') as
+      | AzureDevopsAccounts[]
+      | AzureDevopsError
+  }
+
+  // async getAccessibleResources() {
+  //   return this.get('https://api.atlassian.com/oauth/token/accessible-resources') as
+  //     | AccessibleResource[]
+  //     | AzureDevopsError
+  // }
+
+  async getMyAzureDevopsProfile() {
+    return this.get(
+      `https://app.vssps.visualstudio.com/_apis/profile/profiles/me?api-version=5.1`
+    ) as AzureDevopsUser | AzureDevopsError
+  }
+
+  async getPaginatedProjects(organization: string, url: string, callback: GetProjectsCallback) {
```

Comment:
fetching projects from atlassian's API is pretty clunky, hence the promise + callback.
it looks like azure devops would let us fetch all the projects for all the orgs at once? if so, we can get rid of the callback & just have a chain of promises

Reply:
Similarly to Atlassian, you can fetch projects per organization only. GET https://dev.azure.com/{{myorg}}/_apis/projects?api-version=5.1

I assume we should pick an organization, when a enduser configures the Devops integration.
Does it make sense?

Reply:
it's certainly a tradeoff!
If we force a 1:1 relationship between devops organization & parabol team, then 1 team cannot work with multiple organizations. 

i didn't necessarily love that constraint because teams are often multi-functional and span across different organizations (eg a team includes contractors, who may have their own organization). 

so for atlassian, i just did multiple fetches across every organization that has been added to the team. 80% of the time, it will only be 1 org. for that 20% of the time, the added cost is just an extra http2 request, which IMO is a fair trade to allow a team to work across multiple organizations.

Reply:
True. I agree let's have a similar behavior with Atlassian integration. It's also better from the end-user experience. And yes, the added cost is minimum.